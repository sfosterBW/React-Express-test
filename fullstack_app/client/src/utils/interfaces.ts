//Objects
export interface IFruit {
  id: string
  name: string
  best: boolean
}

export type NewFruit = Omit<IFruit, 'id'>

export interface IFruitResponse {
  status: number
}

//States
export interface AlertState {
  toggle: boolean
}

export interface FruitState {
  data: IFruit[]
}

export interface ModalState {
  toggle: boolean,
  fruit?: IFruit
}
