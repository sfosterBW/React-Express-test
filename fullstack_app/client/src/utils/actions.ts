import { IFruit } from './interfaces'
 
//Action const
export const TOGGLE_ALERT = "TOGGLE_ALERT"

//Action shape
interface ToggleAlert {
  type: typeof TOGGLE_ALERT,
  payload: boolean
}

//Action type
export type AlertActionTypes = ToggleAlert

//Action
export function toggleAlert(alertValue: boolean): AlertActionTypes {
  return {
    type: TOGGLE_ALERT,
    payload: alertValue
  }
}

//Action const
export const TOGGLE_MODAL = "TOGGLE_MODAL"
export const OPEN_MODAL = "OPEN_MODAL"

//Action shape
interface ToggleModal {
  type: typeof TOGGLE_MODAL,
  payload: boolean
}

interface OpenModal {
  type: typeof OPEN_MODAL,
  payload: IFruit
}

//Action type
export type ModalActionTypes = ToggleModal | OpenModal

//Action
export function toggleModal(modalValue: boolean): ModalActionTypes {
  return {
    type: TOGGLE_MODAL,
    payload: modalValue,
  }
}

export function openModal(fruit: IFruit): ModalActionTypes {
  return {
    type: OPEN_MODAL,
    payload: fruit,
  }
}

//Action const
export const GET_FRUITS = 'GET_FRUITS'
export const CREATE_FRUIT = 'CREATE_FRUIT'
export const REMOVE_FRUIT = 'REMOVE_FRUIT'
export const UPDATE_FRUIT = 'UPDATE_FRUIT'

//Action shape
interface GetFruits {
  type: typeof GET_FRUITS,
  payload: IFruit[]
}

interface CreateFruit {
  type: typeof CREATE_FRUIT,
  payload: IFruit
}

interface RemoveFruit {
  type: typeof REMOVE_FRUIT,
  payload: number
}

interface UpdateFruit {
  type: typeof UPDATE_FRUIT,
  payload: IFruit
}

//Action types
export type FruitActionTypes = GetFruits | CreateFruit | RemoveFruit | UpdateFruit

//Actions
export const getFruits = (fruits: IFruit[]): FruitActionTypes => {
  return {
    type: GET_FRUITS,
    payload: fruits
  }
}

export const createFruit = (fruit: IFruit): FruitActionTypes => {
  return {
    type: CREATE_FRUIT,
    payload: fruit,
  }
}

export const removeFruit = (id: number): FruitActionTypes => {
  return {
    type: REMOVE_FRUIT,
    payload: id,
  }
}

export const updateFruit = (fruit: IFruit): FruitActionTypes => {
  return {
    type: UPDATE_FRUIT,
    payload: fruit,
  }
}
