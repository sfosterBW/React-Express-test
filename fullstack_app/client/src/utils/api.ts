import axios from 'axios'
import { IFruit } from './interfaces'
 
axios.defaults.baseURL = '/fruit-api/'

const fetchFruit = async () => {
  const response = await axios.get("list")
  return response.data
}

const createFruit = async (fruit: { name: string, best: boolean }) => {
    const response = await axios.post('new', { new: fruit })
    return response.data
}

const updateFruit = async (fruit: IFruit) => {
    const response = await axios.put('update', fruit)
    return response.data
}

const deleteFruit = async (id: number) => {
  const deleteData = { params: { id: id } }
  const response = await axios.delete('delete', deleteData)
  return response.data
}

export default { fetchFruit, createFruit, updateFruit, deleteFruit }
