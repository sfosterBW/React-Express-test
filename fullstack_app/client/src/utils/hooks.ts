import { useState } from 'react'

export const useField = (
  initValue: string,
  label: string,
  name: string,
  type: string) => {

  const [value, setValue] = useState(initValue)

  const handleChange = (event: any) => {
    setValue(event.target.value)
  }

  const reset = () => {
    setValue("")
  }

  return {
    handleChange,
    label,
    name,
    reset,
    type,
    value
  }
}
