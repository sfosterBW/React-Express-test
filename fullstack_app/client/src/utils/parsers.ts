import { Fruit } from './interfaces'

const isBoolean = (bool: any): bool is boolean => {
  return typeof bool === 'boolean'
}

const isString = (text: any): text is string => {
  return typeof text === 'string' || text instanceof String;
}

const parseBest = (best: any): boolean => {
  if (typeof best === 'undefined' || !isBoolean(best)) {
    throw new Error(`Incorrect or missing best: ${String(best)}`);
  }

  return best
}

export const parseId = (id: any): string => {
  if (!id || !isString(id)) {
    throw new Error(`Incorrect or missing id: ${String(id)}`)
  }

  return id
}

const parseName = (name: any): string => {
  if (!name || !isString(name)) {
    throw new Error(`Incorrect or missing name: ${String(name)}`)
  }

  return name
}

const parseArray = (array: any): any[] => {
  if(array && !Array.isArray(array)) {
    throw new Error(`Incorrect or missing an array ${array}`);
  }

  return array
}

export const toFruit = (object: any): Fruit => {
  return {
    id: parseId(object.id),
    name: parseName(object.name),
    best: parseBest(object.best)
  }
}

export const toFruits = (object: any): Fruit[] => {
  const array = parseArray(object)
  return array.map(fruit => toFruit(fruit))
}

export const toId = (object: any): string => {
  return parseId(object)
}

export default { toFruit, toFruits, toId }
