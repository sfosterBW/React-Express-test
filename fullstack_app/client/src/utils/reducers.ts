import { AlertState, FruitState, ModalState } from './interfaces'
import {
  TOGGLE_ALERT,
  AlertActionTypes,
  TOGGLE_MODAL,
  OPEN_MODAL,
  ModalActionTypes,
  GET_FRUITS,
  CREATE_FRUIT,
  REMOVE_FRUIT,
  UPDATE_FRUIT,
  FruitActionTypes,
} from './actions'

export const initialAlertState: AlertState = {
  toggle: false
}

export function alertReducer(
  state = initialAlertState,
  action: AlertActionTypes): AlertState {
  switch (action.type) {
    case TOGGLE_ALERT: {
      return { ...state, toggle: action.payload || false }
    }
    default:
      return { ...state }
  }
}

export const initialModalState: ModalState = {
  toggle: false,

}

export function modalReducer(
  state = initialModalState,
  action: ModalActionTypes): ModalState {
  switch (action.type) {
    case TOGGLE_MODAL: {
      return { ...state, toggle: action.payload || false }
    }
    case OPEN_MODAL: {
      return { toggle: true, fruit: action.payload }
    }
    default:
      return { ...state }
  }
}

export const initialFruitState: FruitState = {
  data: []
}

export function fruitReducer(
  state = initialFruitState,
  action: FruitActionTypes
): FruitState {
  switch (action.type) {
    case GET_FRUITS: {
      return { ...state, data: action.payload }
    }
    case CREATE_FRUIT: {
      return { ...state, data: state.data.concat(action.payload) }
    }
    case REMOVE_FRUIT: {
      const newFruits = state.data.filter(i => i.id !== action.payload)
      return { ...state, data: newFruits }
    }
    case UPDATE_FRUIT: {
      const updateFruit = state.data.map((fruit) => {
        if (fruit.id === action.payload.id) {
          fruit.best = action.payload.best
          fruit.name = action.payload.name
          return fruit
        } else {
          return fruit
        }
      })
      return { ...state, data: updateFruit }
    }
    default:
      return { ...state }
  }
}
