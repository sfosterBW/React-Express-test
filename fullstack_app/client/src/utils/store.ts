import { applyMiddleware, combineReducers, createStore } from "redux"
import thunkMiddleware from "redux-thunk"
import { composeWithDevTools } from 'redux-devtools-extension'
import {
  alertReducer,
  fruitReducer,
  modalReducer } from './reducers'

export const rootReducer = combineReducers({
  alert: alertReducer,
  fruit: fruitReducer,
  modal: modalReducer
})

export type RootState = ReturnType<typeof rootReducer>

export default function configureStore(preloadedState: undefined) {
  const middlewares = [thunkMiddleware]
  const middlewareEnhancer = applyMiddleware(...middlewares)
  const enhancers = [middlewareEnhancer]
  const composedEnhancers = composeWithDevTools(...enhancers)
  const store = createStore(rootReducer, preloadedState, composedEnhancers)
  return store
}

export const store = configureStore(undefined)
