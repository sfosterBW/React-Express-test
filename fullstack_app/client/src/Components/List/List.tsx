import React, { FC, useState } from 'react'
import { IFruit } from '../../utils/interfaces'
import FruitItem from '../FruitItem/FruitItem'
import styles from './List.module.css'

interface Props {
  fruits: IFruit[]
}

const List: FC<Props> = ({ fruits }) => {

  const type: boolean | undefined = undefined
  const [filter, setFilter] = useState<boolean | undefined>(type)

  const displayFruitList = () => {
    const filterFruit = fruits.filter((i) => filter === i.best || filter === undefined)
    if (filterFruit.length > 0) {
      return filterFruit.map((i) => <FruitItem key={i.id} fruit={i} />)
    } else {
      return <p>Nothing here...</p>
    }
  }

  const setter = (value: boolean | undefined) => () => setFilter(value)

  return (
    <section className={styles.wrapper}>
      <div className={styles.filters}>
        <button className={styles.filter} onClick={setter(undefined)}>All</button>
        <button className={styles.filter} onClick={setter(true)}>Best</button>
        <button className={styles.filter} onClick={setter(false)}>Not best</button>
      </div>
      <div className={styles.devices}>
        {displayFruitList()}
      </div>
    </section>
  )
}

export default List
