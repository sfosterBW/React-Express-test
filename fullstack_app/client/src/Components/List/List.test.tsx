import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import { emptyList, list } from '../../utils/test-helper'

import List from './List'

const mockDispatch = jest.fn()
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch
}))

describe('the List component', () => {
  describe('works with no fruit', () => {
    const component = <List fruits={emptyList} />
    const wrapper = mount(component)

    it('renders with the right structure', () => {
      expect(wrapper).toBeDefined()
      expect(wrapper.find('section')).toHaveLength(1)
      expect(wrapper.find('div')).toHaveLength(2)
      expect(wrapper.find('button')).toHaveLength(3)
      expect(wrapper.find('button').at(0).text()).toEqual("All")
      expect(wrapper.find('button').at(1).text()).toEqual("Best")
      expect(wrapper.find('button').at(2).text()).toEqual("Not best")
    })

    it('renders the same as last time', () => {
      const tree = renderer
        .create(component)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('works with fruit', () => {
    const component = <List fruits={list} />
    const wrapper = mount(component)

    it('renders with the right structure', () => {
      expect(wrapper).toBeDefined()
      expect(wrapper.find('section')).toHaveLength(1)
      expect(wrapper.find('div.row')).toHaveLength(list.length)
      wrapper.find('h3').forEach((title, i) =>
        expect(title.text()).toEqual(list[i].name)
      )
      expect(wrapper.find('button')).toHaveLength((list.length * 2) + 3)
    })

    it('renders the same as last time', () => {
      const tree = renderer
        .create(component)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
