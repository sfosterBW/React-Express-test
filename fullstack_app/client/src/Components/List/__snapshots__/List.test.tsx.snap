// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the List component renders the same as last time 1`] = `
<section
  className="wrapper"
>
  <div
    className="filters"
  >
    <button
      className="filter"
      onClick={[Function]}
    >
      All
    </button>
    <button
      className="filter"
      onClick={[Function]}
    >
      Best
    </button>
    <button
      className="filter"
      onClick={[Function]}
    >
      Not best
    </button>
  </div>
  <div
    className="devices"
  >
    <p>
      Nothing here...
    </p>
  </div>
</section>
`;
